sdk/eventhub/Azure.Messaging.EventHubs/src/Core/ChannelEnumerableSubscription.cs

```diff
@@ -156,28 +142,40 @@ public async ValueTask DisposeAsync()
                 }
                 else
                 {
-                    // TODO (P1//squire): Determine a better approach to prevent a tight loop that consumes resources or
-                    // has excessive allocations.  (see: https://github.com/Azure/azure-sdk-for-net/issues/7094)
-
-                    try
+                    using (waitCancel = CancellationTokenSource.CreateLinkedTokenSource(cancellationToken))
                     {
-                        delayMilliseconds = 50;
+                        if (maximumWaitTime.HasValue)
+                        {
+                            waitCancel.CancelAfter(maximumWaitTime.Value);
+                        }
 
-                        if (waitTime.HasValue)
+                        try
                         {
-                            delayMilliseconds = (int)Math.Min(delayMilliseconds, (waitTime.Value - timer.ElapsedMilliseconds));
+                            // Wait for an item to be available in the channel; if it becomes so, then
+                            // reset the loop so that it can be read and emitted.
+
+                            if (await reader.WaitToReadAsync(waitCancel.Token).ConfigureAwait(false))
+                            {
+                                continue;
+                            }
+                        }
+                        catch (Exception ex) when
+                            (ex is TaskCanceledException || ex is OperationCanceledException)
+                        {
+                            // This is an expected case when the token was canceled or when the channel is
```

Comment:
You should throw if `cancellationToken.IsCancellationRequested`

Reply:
Why would I want to do that?   It's an expected scenario from which we have a viable means to return directly and, therefore, reach the desired end state without an exception path.   I'm not sure that I see value in utilizing an exception as a communications vector for "we have received and honored your request to cancel."   I think the more natural experience is to honor the request and cease iteration cleanly.

Reply:
Because throwing on cancellation in async method is a common pattern in dotnet, this is exactly how other async enumerable would work and is what customers would expect. 

> I don't see value in bubbling cancellation exceptions in this case, as we have a clear path for honoring the request without triggering an exception path.

You are triggering the exception path, you're just choosing to suppress it later.

Reply:
The goal is, of  course, to provide the best experience around consumption.  Cancellation normally triggers an exception when there is no reasonable way to communicate that cancellation was accepted and the lack of concrete understanding could cause confusion to callers.   There is no common practice of which I'm aware for triggering exceptions on cancellation as a general rule.   Obviously, the intent is not to go against expectations or de facto standards, however.

Got any references that you can point me at around other async enumerables that aren't bound to a finite set where there could be confusion as to "am I out of things or did cancellation go through?"   I'm open to considering it, but I feel like we're comparing apples to oranges here.

Reply:
>  There is no common practice of which I'm aware for triggering exceptions on cancellation as a general rule. 

But there is, in BCL async methods throw cancellation exceptions if cancellation token gets canceled during their execution (modulo race conditions of course).

> Got any references that you can point me at around other async enumerables that aren't bound to a finite set or where there wouldn't be be confusion as to "am I out of things or did cancellation go through?" I'm open to considering it, but I feel like we're comparing apples to oranges here.

My point is that customer might not even realize that the enumerable is infinite and they have to expect different behavior from it. 

So they would write something like:

```
Task<int> Summ20EventValues(CancellationToken ct)
{
    int sum = 0;
    int i = 0;
    async foreach(var event in channel.SubscribeToEvents(ct))
    {
        sum += event.Properties["Value"];
        i++;
        if (i>=20)
        {
            break;
        }
    }

    return sum;
}
```

And expect to get an exception in case of cancellation, but the method would successfully complete returning the wrong value instead.

Reply:
Fair enough.   I think you've convinced me on the merits in this case.  Still feels like using exceptions as flow control, but your through process makes sense.